File: 264.丑数-ii.go
/*
 * @lc app=leetcode.cn id=264 lang=golang
 *
 * [264] 丑数 II
 */

// @lc code=start
func nthUglyNumber(n int) int {
	res := []int{1}
	index2,index3,index5 := 0,0,0
	min := func(x,y int) int {
		if x < y {
			return x
		}
		return y
	}
	
	for {
		if len(res) == n {
			return res[n-1]
		}
		next := min(2*res[index2],min(3*res[index3],5*res[index5]))
		res = append(res,next)
		// 下面不应该用不等号<=，而应该用==，因为不可能出现<的情况
		for 2*res[index2] <= next {
			index2++
		}
		for 3*res[index3] <= next {
			index3++
		}
		for 5*res[index5] <= next {
			index5++
		}
	}
}
// @lc code=end

File: 264.丑数-ii.ts
/*
 * @lc app=leetcode.cn id=264 lang=typescript
 *
 * [264] 丑数 II
 */

// @lc code=start
function nthUglyNumber(n: number): number {
    let res: number[] = [1];
    let i2 = 0, i3 = 0, i5 = 0;
    for (let i = 1; i < n; i++) {
        res[i] = Math.min(res[i2] * 2, res[i3] * 3, res[i5] * 5);
        if (res[i] === res[i2] * 2) {
            i2++;
        }
        if (res[i] === res[i3] * 3) {
            i3++;
        }
        if (res[i] === res[i5] * 5) {
            i5++;
        }
    }
    return res[n-1];
};
// @lc code=end

